List package "files" inside package.json

Hookyns opened the pull request:
This optimizes the size of the published package. Currently, there is the whole repo published to npm.

Hookyns commented:
OT:
Hello @spion,
Thanks for this repo. I wanted to create Solid.js wrapper for Ionic few weeks ago, but I had no time for that and now I found this repo you've "just" created. 😆 

What's your intention and progress so far? Are you open to collaboration?

spion commented:
Hey @Hookyns - thanks for the contribution!

The wrapper works, although the router wrapper is very basic. Its currently serving my needs on two projects, but I know that this is not necessarily sufficient for apps that have things like tab navigation. My intention has been to start matching the capabilities described in https://ionicframework.com/docs/api/router (the stencil / webcomponent version) by reusing as much of the existing router component as possible - but this has been slow as I don't currently need those features.

One particular thing I'm not 100% happy about is the mixed casing. Components are right now named the same as in React; however their attributes still use the `snake-case` which is characteristic of the underlying webcomponents. So you get things like

```tsx
<IonAlert message="Hi" sub-header="Hello World" />
```

Events on the other hand continue to be automatically camelCased so things like `on:ionChange` are correct.

I already have quite a bit of code using this library :sweat_smile: but I think I'd be willing to change it if there is a good idea on how to generate a performat wrappers with solid-idiomatic attribute naming

For the published library's size, a lot of it may be due to the fact that it republishes ionicons as well. Currently, coarse-grained tree-shaking for this should work (if you don't use icons you don't pay for it), but I think it doesn't work if you only want to use a few icons rather than the entire set. This is something that could probably use some improvement as well (splitting it out into a separate package seems like a good first step).

What are your thoughts? Where would you like to see the library move?

Hookyns commented:
Well, I would start from the repo itself. 
- I'd like to turn it into monorepo (I usually use **pnpm** workspaces but **bun** has workspaces too so no problem). Then you can reference packages inside repo by using eg.: `"ionic-solidjs": "workspace:*"`.

  > I found you use "file:../" inside `/example`  to reference the main package but it does not work with **bun** (at least not on my machine), so I expect you used `npm` for that example?! 

  To allow using the main package (ionic-solidjs) inside a workspace I have to move it to some subfolder (probably `/packages/ionic-solidjs`).

- I'd like to introduce `biome` instead of prettier, so there is also linting, etc.

- As you mentioned Icons, I would like to separate icons into a standalone package. `IonIcon` should be part of the core, but icons themself shouldn't IMO. You can still use SVGs directly `import {someIcon} from 'ionicon'; <IonIcon icon={someIcon} />` or use custom icons. So we can introduce eg. `@ionic-solidjs/icons` (_and maybe publish `@ionic-solidjs/core` instead of direct `ionic-solidjs` to keep some kind of package naming convention? I'm not sure about this one. What do you mean?_)

---
Other things
- I changed the ionic config initialization (which you do inside every component) to [platform config](https://ionicframework.com/docs/developing/config#per-platform-config), the same way eg. React implementation does it.
- I had to fix noShadowDOM() inside `SolidIonRouteAcceptor`; it was throwing errors so I assigned renderRoot directly (exactly the same way the noShadowDOM function does it). But I'm not sure why it is here, inside SolidIonRouteAcceptor. Shouldn't it be optional, added by a user if needed? Do you have a repro with that modal case?
- Routing and Tabs, yeah,... I tried to use ionic-solidjs with generated ionic start app for React. I rewrote a few lines and I got stuck on Tabs. React uses the React router, not the Ionic router so you cannot use Ionic's React docs; you have to follow docs for web components, but you made a custom implementation of the Router. It's not IonRouter (it breaks the naming convention) and the implementation is different (maybe that's why you changed the name). So developers cannot follow either documentation.

  I am thinking about implementing IonSolidRouter, the same way Ionic did it for React. So developers can stick to routing they know from Solid. Or they can use IonRouter which will be the basic wrapper of `<ion-router>`.
- Attribute naming, I noticed... I'll try to figure something out.

spion commented:
My thoughts:

- re: workspaces, agreed that would be the logical step when separating the icons
- re: linter, I'm not a huge fan of JS linters in general as a lot of them come with default rules that are sometimes just plain wrong, but biome as a prettier replacement sounds great!
- Agreed on everything regarding icons. Wondering if the package namespace should have "community" to signify its not "official"? e.g. `@ionic-solidjs-community/pkgname`

For the other bits:

- Thanks for the platform config change. It looks like I haven't been following the recommendations.
- I don't have a specific code repro but I could add a modal with trigger="trigger-element" to the example. The trigger element wont work if shadow dom is enabled.

Regarding the custom router, its not very ergonomic to use webcomponent IonRouter directly. As far as I can tell, that router seems to only works with registered webcomponent string names, so each SolidJS page component would need to be defined as a web-component. The idea behind the router is to therefore be a thin wrapper on top of IonRouter where a single component (route acceptor) re-routes to the right Solid component and passes down the correct context. (In fact, passing user-defined context through is something that might not even work at the moment). 


I think IonSolidRouter might be quite difficult to do well - its the main source of complication for the Ionic team to implement new frameworks. In theory, there is nothing preventing any ionic app from using the standard Solid router - the only thing missing would be transitions. Its also not clear to me how the ionic nested routing behavior (separate history per subtree) would map to solid's router. A separate package to explore those ideas seems reasonable.

For this simple router's name, the name change to IonRouter seems OK, but I think only if the API starts to look more like the one seen in Ionic - right now the current router is simply a different, ultra-simple API that "gets the job done". This could also go into a separate package. It might make sense to have multiple router implementation explorations in general :grin:

Hookyns commented:
@spion are you OK with TABs over spaces (so everybody can set prefered tab width in their IDE)? 😄 

---

BTW: I probably figured out the props/attributes. Ionic uses Stencil and all the JS properties on the Components (custom HTML elements) are decorated by `@Prop` and it detects changes and rerender in case the prop is changed. So we can use properties directly, not the HTML attributes. Properties should be faster than Attributes and there is no issue with naming because names of those properties match the properties in the exported TS interfaces. We just have to prefix properties by `prop:` and Solid will handle it. Do you think this is OK?

> src/utils/fixProps.ts
```typescript
const EVENT_PROP_REGEX = /^on[A-Z]/;

export function fixProps(props: Record<string, any>) {
    const newProps: Record<string, any> = {};
    const keys = Object.keys(props);

    for (const key of keys) {
        if (key.includes(':') || key === 'children' || EVENT_PROP_REGEX.test(key)) {
            newProps[key] = props[key];
        }
        else {
            newProps[`prop:${key}`] = props[key];
        }
    }

    return newProps;
}
```

> Generated component wrapper
```typescript
export function IonAlert(props: IonAlertProps) {
  const fixedProps = () => fixProps(props);
  //@ts-ignore
  return <ion-alert {...fixedProps()} />
}
```

But when we'll use props instead of attributes you will not be able to see values in the DOM, idk if it's problem or not...
![image](https://github.com/spion/ionic-solidjs/assets/2551259/2b67972f-11fb-4010-be68-3da3ac168f65)
> Neither header nor sub-header are in the DOM.

Usage
![image](https://github.com/spion/ionic-solidjs/assets/2551259/88c4a285-8f34-42c8-b4ac-660bdfaecb11)

spion commented:
Tabs :imp: nooo... Actually as long as I don't have to think about it, its fine

Regarding props, nice find. I think this is OK. Hopefully solid-devtools will eventually... solidify :grinning:

Hookyns commented:
> Tabs 👿 nooo... Actually as long as I don't have to think about it, its fine

Biome should handle it, you just have to install plugin so the editor can invoke it on save.
But your repo, your call. 😄 You can change it back to spaces whenever you want. 👍 This codebase is small so I can survive that 2 space indentation. 

I'll enable Biome's linter too, but with few rules disabled or changed to warns,.. my regular config. Feel free to change the configuration.

---

Should I create PR to main or should I target different branch so you can test it first?

spion commented:
No worries - I should be able to pull the branch with `gh pr checkout`.

Regarding the tabs, yeah as long as format on save does the correct thing I'm completely fine with it it.

However, I would strongly prefer continuing to use `let` instead of `const` camp which may be one of my more controversial linter stances. I find `const` significantly impares my ability to read the code and has a subtly misleading effect on my thinking (even after trying to get used to it). So this would probably be the one rule I wouldn't want turned on: https://biomejs.dev/linter/rules/use-const/ - hope that's not a deal-breaker!

Hookyns commented:
> However, I would strongly prefer continuing to use `let` instead of `const` camp which may be one of my more controversial linter stances. I find `const` significantly impares my ability to read the code and has a subtly misleading effect on my thinking (even after trying to get used to it). So this would probably be the one rule I wouldn't want turned on: https://biomejs.dev/linter/rules/use-const/ - hope that's not a deal-breaker!

Oh, interesting. I've met nobody with sutch stance. Well, I've already rewrote all `let`s that should be const to `const`s. :/
We can turn that lint option off, but I think I'll not be able to start using `let` for everything,.. consts are already hardcoded in my brain. But I have no issue to keep `let`s where you put them.


---

Here is a branch with reworker repository - monorepo and all the changes we talked about. 
https://github.com/Hookyns/ionic-solidjs/tree/features/monorepo-reworked

Regarding the routing, I created wrappers IonRouter and IonRoute, so inside `@ionic-solidjs/core` there is whole Ionic routing system. I moved your Router to `@ionic-solidjs/router` (I used new IonRouter and IonRoute components in the implementation).
